pyproject.toml

```diff
@@ -260,6 +260,49 @@ exclude= [
 ]
 
 
+[tool.mypy]
+python_version = "3.10"
+files = ["astropy/**/*.py"]
+plugins = ["numpy.typing.mypy_plugin"]
+ignore_missing_imports = true
+follow_imports = "skip"
+show_error_codes = true
+pretty = true
+
+    [[tool.mypy.overrides]]
+     module = [
+        "astropy/logger",
+        "astropy/_erfa/*",
+        "astropy/config/*",
+        "astropy/conftest",
+        "astropy/constants/*",
+        "astropy/convolution/*",
+        "astropy/coordinates/*",
+        "astropy/cosmology/*",
+        "astropy/extern/*",
+        "astropy/io/*",
+        "astropy/modeling/*",
+        "astropy/nddata/*",
+        "astropy/samp/*",
+        "astropy/stats/*",
+        "astropy/table/*",
+        "astropy/tests/*",
+        "astropy/time/*",
+        "astropy/timeseries/*",
+        "astropy/uncertainty/*",
+        "astropy/units/*",
+        "astropy/utils/*",
+        "astropy/version",
+        "astropy/visualization/*",
+        "astropy/wcs/*",
+     ]
+     ignore_errors = true
```

Comment:
Very few maintainers have expressed any interest in adding type annotations and the number of people that have expressed interest in type checking `astropy` is even lower. It seems better to require sub-packages to explicitly opt in to type checking instead of setting `files = ["astropy/**/*.py"]` and then explicitly opting sub-packages out.

Reply:
A consideration here is whether we want subpackages that get added to Astropy in the future to be type checked with mypy by default.  Because adding type hints is significantly easier when writing code for the first time, my inclination would be to set it so that new subpackages do get type checked by default. But, this consideration would only matter if any subpackages might be added in the next few years.

Reply:
> A consideration here is whether we want subpackages that get added to Astropy in the future to be type checked with mypy by default.

Unlikely we will have a new subpackage anytime soon (but never say never). When we do though, it is usually already an existing code/package. To require typing right off the bet would be a deal breaker. So 👎 from me.

Reply:
> Because adding type hints is significantly easier when writing code for the first time, my inclination would be to set it so that new subpackages do get type checked by default.

It is true that type checking is simpler if it is done from the very beginning, but we can't demand that from new sub-packages without an APE.

Reply:
Agreed that new sub-packages in principle new sub packages can be more easily typed, but in practice are unlikely to be given their origins from other sources.

> Very few maintainers have expressed any interest in adding type annotations and the number of people that have expressed interest in type checking astropy is even lower.  It seems better to require sub-packages to explicitly opt in to type checking ...

I think the picture is more complicated. It's true that few maintainers have wanted to add type annotations and fewer have started, but what's also true is that many users want it. (I have plenty of colleagues that have told me that they struggle to use Astropy for a few reasons and for many of them a significant one is because they aren't overly interested in programming and just want to get the job done and have justifiably gotten really used to the way that modern IDEs can surface information from type annotations and help do argument-by-argument completion of functions and classes. I know that I'm super used to it and am less-than-thrilled when I have to use a library that makes me dig around for the information. I'm looking at you `jax.experimental.array_api`). The bar for good, helpful user-facing code has been raised. So while it's true that few maintainers are doing anything, I think that reflects a failing on our part, not a reason to structure Astropy's config.

That being said, in its current form this PR is exactly equivalent to not opting in for any sub-package. And it will basically remain so for a while. The crucial difference between adding an opt-in list versus an opt-out list is that an opt-out list is a visible reminder of what remains to be done for Astropy to be more performant, have fewer bugs, and be more user friendly. I'm hoping that `utils` and `cosmology` and `units` will be some of the first libraries to remove themself from the opt-out list. I know users will appreciate it. 

@eerovaher you are doing an excellent job adding type hints and other than my thanks right now, it's probably a mostly thankless task. Thank you for your work on `utils`.

Reply:
We should indeed add type annotations that are helpful for IDE tab-completion, but that does not require configuring a type checker.

Ideally type checking would be helpful for developers, but in practice whenever I try to use `mypy` in `astropy` I immediately get overwhelmed by warnings from `units` and that is not helpful.

Reply:
> We should indeed add type annotations that are helpful for IDE tab-completion, but that does not require configuring a type checker.

We do need to test the type annotations. RN the ones we are adding are completely untested. They look right...

Reply:
The complexity and rigor of annotations that are required to satisfy type checkers is much higher than what can be useful for IDE tab-completion. I know that some of the annotations I've written are oversimplified so that a type checker will (rightfully) complain about them, but I still wrote them that way because having good return type annotations is helpful for IDEs even if some of the parameter annotations are sloppy. Furthermore, my expectation is that it will be simpler to start using a type checker when the fraction of annotated code is larger because then the checker should be able to point out more precisely what needs to be changed to make it happy.

I don't want us to end up in a situation where the few people that are adding annotations have to spend more time adding exceptions to the type checker configuration than they would spend on adding annotations themselves. And if early on our approach is to ignore (almost) every type violation then we can do that without writing anything down.

Reply:
I feel like type hints too often take 100% of our focus in these discussions, when in fact they're not the *only* reason to use a type checker, so let me give a different perspective. Personally, the main reason I'd want to see astropy starting to use a type checker is that it'd enable using constructs from the `typing` module that are essentially declarations of intention which don't have any effect at runtime, like `final`, `Final`, `override`, `overload`, `assert_never`[^1], `Protocol` or `ReadOnly` (new in Python 3.13). To me, these are the real game changers and they are the part the typing language that actually allows me to detect and fix defects through a type checker, but of course type hints play a big part in enabling this too. I'm not saying that type hints don't help finding bugs too (they do !), but in my experience they're higher cost, lower rewards in that area. 

[^1]: this one is not a no-op at runtime

Reply:
@neutrinoceros, a good point! I use `Protocol` all the time, and `TypedDict`, and other useful runtime-features. `override` is also such a nice addition to `typing`.